src/topology/config/watcher.rs

```diff
@@ -0,0 +1,125 @@
+use super::Config;
+use notify::{watcher, DebouncedEvent, RecommendedWatcher, RecursiveMode, Watcher};
+use std::{
+    path::{Path, PathBuf},
+    sync::mpsc::{channel, Receiver},
+    thread,
+    time::Duration,
+};
+
+/// Per notify own documentation, it's advised to have delay of more than 30 sec,
+/// so to avoid receiving repetitions of previous events on macOS.
+///
+/// But, config and topology reload logic can handle:
+///  - Invalid config, caused either by user or by data race.
+///  - Frequent changes, caused by user/editor modifying/saving file in small chunks.
+/// so we can use smaller, more responsive delay.
+pub const CONFIG_WATCH_DELAY: std::time::Duration = std::time::Duration::from_secs(1);
+
+const RETRY_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
+
+/// On unix triggers SIGHUP when file on config_path changes.
+/// Accumulates file changes until no change for given duration has occured.
+/// Has best effort guarante of detecting all file changes from the layend of
+/// this function until the main thread stops.
+///
+/// Doesn't do anything on Windows.
+pub fn config_watcher(config: &Config, config_path: PathBuf, delay: Duration) {
+    if config.global.reload_config {
+        #[cfg(unix)]
+        {
+            use nix::sys::signal;
+            // Create watcher now so not to miss any changes happening between
+            // returning from this function and thread started.
+            let mut watcher = create_watcher(&config_path, delay);
+
+            thread::spawn(move || loop {
+                if let Some((_, receiver)) = watcher.take() {
+                    info!("Watching configuration file");
+                    while let Ok(msg) = receiver.recv() {
+                        match msg {
+                            DebouncedEvent::Write(_)
+                            | DebouncedEvent::Create(_)
+                            | DebouncedEvent::Remove(_) => {
```

Comment:
I have not tested this code yet, but would it make sense to include here the last time we got a config update that we want, and use that as the timer filter instead of using a thread sleep and a watch delay? Not sure what the implications of that might be but I assume it might make this logic a bit more clear and easier to reason about.

Reply:
If I understood you right, you mean implementing our own delay implementation and use it with immediate watch. I think it would be more complex, and even then we won't be able to avoid the `thread::sleep(RETRY_TIMEOUT);`, because it's here to backoff for a few seconds when watcher errors for some reason.

Reply:
Right, what Im saying is we should limit the time to get the next event to say 1second, then within our code use a instant to determine if we have waited long enough to avoid false positives. So instead of pushing that timeout logic to the watcher, I think it would make sense to instead implement that time check within our code so its clear that that is the behavior.

Reply:
Well, code of that turned out simpler than expected.